### DISCUSSION TITLE
Add contextmanager to engine

### DISCUSSION THREAD

**Discussion by akazantsev-anna-money** [sqlalchemy/sqlalchemy#7380] [Ideas]
After use engine need to call `dispose()` method. Now when I use a framework (for exmple `aiohttp`) I wrote something like
```
def setup_db(app: web.Application):
    from sqlalchemy.ext.asyncio import create_async_engine

    engine = create_async_engine(
        get_connection_url(),
        pool_recycle=300,
    )
    app["db"] = engine

    async def pg_engine(app: web.Application):
        try:
            yield
        finally:
            await app["db"].dispose()

    app.cleanup_ctx.append(pg_engine)
```

If engene have a context manager then code be more readable in pg_engine function
```
    async def pg_engine(app: web.Application):
        async with app["db"]:
            yield
```

**Comment by CaselIT** [comment]
Hi,

I'm personally -1 on this, since it would encourage sub-optimal patterns.
In case of a web application the engine should *not* be re-created at each requests as you appear to be doing.
Unless there is a particular reason, a single engine should be shared by all request.
Usually for a web application the following pattern is used:
- the engine is initialized at the module level or alternatively on application startup
- for each request a single connection/session is created
- at the end of the request the connection/session is closed

**Reply by heckad** [reply]
[app.cleanup_ctx](https://docs.aiohttp.org/en/stable/web_advanced.html#aiohttp-web-cleanup-ctx) executes only ones.

**Reply by CaselIT** [reply]
Still it does not seem to be too cumbersome to use their example:
```py
async def setup_db(app):
    from sqlalchemy.ext.asyncio import create_async_engine
    app['db'] = create_async_engine(
        user='postgre',
        database='postgre',
        host='localhost',
        port=5432,
        password=''
    )
    yield
    await app['db'].dispose()

app.cleanup_ctx.append(setup_db)
```

**Comment by zzzeek** [comment]
I'll add to @CaselIT 's answer, SQLAlchemy has these two separate concepts:

* **factories** - engines, sessionmaker
* **resources** - Connection, Session

Resources are a **per request** kind of object - these should be **short lived, function scope**. For these, closing them out is important and necessary for each use.   

A factory OTOH is **per application** (or perhaps per process).  These should be **long lived, module scope**.  There is normally no reason to "dispose" of a factory.  The Engine has a `.dispose()` method since it houses a connection pool, the cases where this method is appropriate are:
* when moving over process boundaries, like using os.fork() or multiprocessing
* in test suites, for suites that connect to a particular database and want to clean up when the suite is finished
* *maybe* in a shell-script / CLI / cron job kind of application, if the database driver in use does not seem to clean up connections gracefully when the application exits, dispose() can be used at application end to close off connections (though you probably should just use NullPool)

and that's it!  engine.dispose() doesn't belong anywhere in a web application.
